### PR TITLE
Feature/bloom filter

### PR DESCRIPTION
**Pull request by keyseven123**
## Purpose of the Change and Brief Change Log
  This pull request adds an optional BloomFilter to our chained hashmap. This helps not having to traverse potentially long chains.
Additionally, we provide a [nautilus fix of right shift](https://github.com/nebulastream/nautilus/pull/291). This fix was found by @LucaCode37

On my local machine, I saw speedups of up to 40x for the Hashjoin assuming a best case scenario for the bloom filter of one long chain.
  
  ## Verifying this change
  This change is tested by running our existing tests.
  
  ## What components does this pull request potentially affect?
  - Dependencies — adds `rapidcheck` (vcpkg + nix flake)
  - ExecutionEngine — stream join `Build` / `Probe` operators, slice store
  - QueryCompiler — `LowerToPhysicalHashJoin` / `LowerToPhysicalNLJoin` + new `LowerToJoinUtil`
  - nes-nautilus (new `BloomFilterRef`)
  
  ## Documentation
  - The change is reflected in the necessary documentation, e.g., design document, in-code documentation.
  - All necessary methods (no getter, setter, or constructor) have a documentation.

### CODE REVIEW

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+    if (expectedEntries == 0)
+    {
+        return {.bitCount = 64, .hashCount = 1};
+    }
+    const auto n = static_cast<double>(expectedEntries);
+    const double ln2 = std::numbers::ln2;
+    const auto m = std::max(static_cast<uint64_t>(std::ceil(-n * std::log(fpRate) / (ln2 * ln2))), uint64_t{64});
+    const auto k = std::max(static_cast<uint64_t>(std::round(static_cast<double>(m) / n * ln2)), uint64_t{1});
+    return {.bitCount = m, .hashCount = k};
+}
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: why is this function here? it should be part of the bloom filter.

**Review by ls-1801 — CHANGES_REQUESTED** [review]
I think that, in the current state, adding this level of complexity does not outweigh the benefits for me. Especially with long-awaited changes from @ThanGeo on the horizon, I would really like to postpone this.

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp (hunk 1 above)
Comment (keyseven123) [reply nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: Done

**Comment by keyseven123** [comment]
> I think that, in the current state, adding this level of complexity does not outweigh the benefits for me. Especially with long-awaited changes from @ThanGeo on the horizon, I would really like to postpone this.

I have rebased and stacked this pr on top of the chained hash map refactoring pr

**Review by ls-1801 — CHANGES_REQUESTED** [review]
re-request once relevant

**Comment by keyseven123** [comment]
> re-request once relevant

I have rebased this pr/branch on the new chained hash map. Please let me know what you guys think

File: nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    return BloomFilterRef{NoOp{}};
+}
+
+BloomFilterRef BloomFilterRef::createDisabled()
+{
+    return BloomFilterRef{NoOp{}};
+}
+
+BloomFilterRef::Impl::Impl(const BloomFilterParams params, std::unique_ptr<HashFunction> hashFunction)
```
Comment (ThanGeo) [inline nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp]: Is this Impl::Impl really the best way? Kinda reduces readability

File: nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    static BloomFilterRef createDisabled();
+
+    [[nodiscard]] const BloomFilterParams& params() const { return bloomFilterParams; }
+
+    void add(const nautilus::val<uint64_t*>& bitsData, const HashFunction::HashValue& hash) const
+    {
+        std::visit([&](const auto& bloomFilter) { bloomFilter.add(bitsData, hash); }, impl);
+    }
+
+    [[nodiscard]] nautilus::val<bool> mightContain(const nautilus::val<uint64_t*>& bitsData, const HashFunction::HashValue& hash) const
```
Comment (ThanGeo) [inline nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp]: what does "mightContain" mean?

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        return 0.0;
+    }
+    const auto exponent = -static_cast<double>(numHashes) * static_cast<double>(numEntries) / static_cast<double>(numBits);
+    return std::pow(1.0 - std::exp(exponent), static_cast<double>(numHashes));
+}
+
+/// Wraps the Nautilus-compiled add/mightContain pair. Engine setup and function registration happen
+/// once per instance so the hot loop is just function-call overhead. The `enabled` flag picks
+/// between the real BloomFilter variant and the no-op variant
+class TestableBloomFilter
```
Comment (ThanGeo) [inline nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: Following the changes to TestableChainedHashMap and TestablePagedVector being moved to their own class files, maybe consider doing the same here.

**Review by ThanGeo — CHANGES_REQUESTED** [review]
Overall minor changes requested. One question, are you using a single hash function in the bloom filter? By definition usually $K$ hash functions are used even though with added overhead but better correctness

File: nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp (hunk 3 above)
Comment (keyseven123) [reply nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp]: If a hash/key might be seen by the bloom filter.

File: nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp (hunk 2 above)
Comment (keyseven123) [reply nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp]: Renamed it to `Active/Inactive`.

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp (hunk 4 above)
Comment (keyseven123) [reply nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: Done

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 5]
```diff
@@ -346,4 +376,31 @@ TEST(ChainedHashMapIteratorTest, emptyMapIsAnEmptyRange)
...
+    putAndLookupKeysProperty(TestUtils::EngineMode::Interpreter, true);
+}
+
+RC_GTEST_PROP(ChainedHashMapPropertyTest, putAndGetAllWithBloomFilterCompiler, ())
+{
+    Logger::setupLogging("ChainedHashMapPropertyTest.log", LogLevel::LOG_DEBUG);
+    putAndGetAllProperty(TestUtils::EngineMode::Compiler, true);
+}
+
+RC_GTEST_PROP(ChainedHashMapPropertyTest, putAndGetAllWithBloomFilterInterpreter, ())
```
Comment (ThanGeo) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:504]: Maybe add one more property test that check correctness: e.g. 2 chained hash maps, one with bloom filter and one with out, results should always match. Also maybe tests with very high/very low collision on the bloom filter? I',m trying to think of tests that will reveal instantly if any future changes break any edge cases or something

**Review by ThanGeo — CHANGES_REQUESTED** [review]

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp (hunk 5 above)
Comment (keyseven123) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:504]: Good idea. Added tests for it.

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -75,7 +75,7 @@ jobs:
             echo "=== sccache config before build ==="
             sccache --start-server
             sccache --show-stats
-            cmake -GNinja -B build -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} ${{ matrix.cmake_flags }}
+            cmake -GNinja -B build -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} -DNES_EXHAUSTIVE_TESTS=1 ${{ matrix.cmake_flags }}
```
Comment (ls-1801) [inline .github/workflows/build_and_test.yml]: The whole point was to not enable exhaustive tests by default; we can run them in the nightly.
But i would manage that via the matrix.yaml

File: nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    /// Zeroes `dst`. Does nothing when the filter is Inactive.
+    void clearBits(const nautilus::val<std::byte*>& dst) const
+    {
+        std::visit([&](const auto& bloomFilter) { bloomFilter.clearBits(dst); }, filter);
+    }
+
+private:
+    /// Real BloomFilter implementation. Method bodies live in BloomFilterRef.cpp.
+    class Active
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp]: In general, I like the idea of being able to switch off the Bloom filter. But I think this should be a decision made by the Bloom filter owner. So the chained hash map can be configured to use a BF or not, and the implementation in the chained hash map should be conditional. 

I think it's kinda odd to have a Bloom filter that does nothing. I would be explicit and have a data structure that doesn't use a Bloom filter.

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp [hunk 8]
```diff
@@ -73,8 +73,16 @@ class ChainedHashMap final : public HashMap
...
+    /// @param bloomFilterMemAreaSize bytes for the optional in-map BloomFilter bit area, 0 disables it (see getBloomFilterMemArea)
+    static void
+    init(TupleBuffer& tupleBuffer, uint64_t entrySize, uint64_t numberOfBuckets, uint64_t pageSize, uint64_t bloomFilterMemAreaSize);
+    static void init(
+        TupleBuffer& tupleBuffer,
+        uint64_t keySize,
+        uint64_t valueSize,
+        uint64_t numberOfBuckets,
+        uint64_t pageSize,
+        uint64_t bloomFilterMemAreaSize);
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp]: At this point, I would use a ChainedHashMapConfig object, which can contain a std::optional<BloomFilterConfig>

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp [hunk 9]
```diff
@@ -105,10 +113,19 @@ class ChainedHashMap final : public HashMap
 
     [[nodiscard]] uint64_t getMask() const { return header().mask; }
 
+    /// Size in bytes of the optional in-map BloomFilter bit area (0 when disabled). Used to propagate the
+    /// filter sizing when a derived map is created from an existing one (e.g. the aggregation final map).
+    [[nodiscard]] uint64_t getBloomFilterMemAreaSize() const { return header().bloomFilterMemAreaSize; }
+
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp]: do we need to expose that?

File: nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    /// Each iteration computes a derived hash position via h_i(x) = (lower32 + i * upper32) mod bitCount.
+    /// The position is split into a word index (which uint64_t in the array) and a bit offset within that word.
+    /// Iterating hashCount times sets multiple independent bits, reducing the false-positive rate.
+    for (nautilus::val<uint64_t> i{0}; i < params.hashCount; i = i + nautilus::val<uint64_t>{1})
+    {
+        nautilus::val<uint64_t> x = (lower32 + i * upper32) % params.bitCount;
+        const nautilus::val<uint64_t> wordIndex = x / nautilus::val<uint64_t>{64ULL};
+        nautilus::val<uint64_t> bitOffset = x % nautilus::val<uint64_t>{64ULL};
+        const nautilus::val<uint64_t> mask = nautilus::val<uint64_t>{1ULL} << bitOffset;
+        *(bitsMemoryRef + wordIndex) = *(bitsMemoryRef + wordIndex) | mask;
```
Comment (ls-1801) [inline nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp]: I think the math for the setup is off. You divide by 8, which means for 65 bits we would use 9 bytes. However, here you would access 16 bytes. I think all your tests pad to 64 bits, so this wouldn't happen, but as soon as someone doesn't give a multiple of 64, this could crash.

File: nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+void BloomFilterRef::Active::add(const nautilus::val<uint64_t*>& bitsMemoryRef, const HashFunction::HashValue& hash) const
+{
+    nautilus::val<uint64_t> lower32 = hash & nautilus::val<uint64_t>{(1ULL << 32) - 1};
+    nautilus::val<uint64_t> upper32 = hash >> nautilus::val<uint64_t>{32ULL};
+
+    /// Each iteration computes a derived hash position via h_i(x) = (lower32 + i * upper32) mod bitCount.
+    /// The position is split into a word index (which uint64_t in the array) and a bit offset within that word.
+    /// Iterating hashCount times sets multiple independent bits, reducing the false-positive rate.
+    for (nautilus::val<uint64_t> i{0}; i < params.hashCount; i = i + nautilus::val<uint64_t>{1})
```
Comment (ls-1801) [inline nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp]: is there a reason this is not a static_val?

**Review by ThanGeo — APPROVED** [review]

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+/// input - including numKeys = 0. Covers the old `edgeCaseZeroExpectedEntries` and `gettersReturnReasonableValues`.
+void sizingProperty()
+{
+    const auto numKeys = *rc::gen::inRange<uint64_t>(0, MAX_NUM_KEYS_SIZING + 1);
+    const auto fpRate = *rc::gen::elementOf(FP_RATES);
+    const auto params = BloomFilterParams::compute(numKeys, fpRate);
+    RC_ASSERT(params.bitCount >= 64);
+    RC_ASSERT(params.hashCount >= 1);
+    RC_ASSERT(params.allocationByteCount() >= 8);
+}
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/BloomFilterTest.cpp:288]: i think this is kinda pointless to test here. Especially since we don't really look at interesting large integer values where overflow could become an issue.

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+namespace NES::Nautilus::Interface
+{
+namespace
+{
+
+using TestUtils::EngineMode;
+using TestUtils::TestableBloomFilter;
+
+/// Configured FP rates drawn per property.
+constexpr std::array<double, 6> FP_RATES = {0.0001, 0.001, 0.01, 0.05, 0.1, 0.2};
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: why limit our self to these? couldnt we just pass a range from 0.0001 to 0.2?

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+/// reported present — and the measured false-positive rate stays within the theoretical bound for the
+/// actual saturation. Single-threaded: one filter, one bit array.
+void insertAndCheckProperty(EngineMode mode)
+{
+    const auto numKeysForSizing = *rc::gen::inRange<uint64_t>(MIN_NUM_KEYS_SIZING, MAX_NUM_KEYS_SIZING + 1);
+    const auto fpRate = *rc::gen::elementOf(FP_RATES);
+    const auto numUnique = *rc::gen::inRange<uint64_t>(MIN_NUM_UNIQUE, MAX_NUM_UNIQUE + 1);
+    const auto totalMultiplier = *rc::gen::inRange<uint64_t>(MIN_TOTAL_MULTIPLIER, MAX_TOTAL_MULTIPLIER + 1);
+    const auto numTotal = numUnique * totalMultiplier;
+    const auto seed = *rc::gen::arbitrary<uint64_t>();
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: You probably want to wrap this in a `rc::gen::noShrink()`

rc would attempt to find a minimal reproducible test case, however, shrinking the seed does not have any useful benefit

File: nes-physical-operators/include/HashMapOptions.hpp [hunk 15]
```diff
@@ -42,7 +43,8 @@ struct HashMapOptions
         const uint64_t keySize,
         const uint64_t valueSize,
         const uint64_t pageSize,
-        const uint64_t numberOfBuckets)
+        const uint64_t numberOfBuckets,
+        Nautilus::Interface::BloomFilterRef bloomFilter)
```
Comment (ls-1801) [inline nes-physical-operators/include/HashMapOptions.hpp]: Passing a bloom filter Ref in the HashMapOptions seems kinda odd. Probably saner to have the data structure own the BloomFilterRef and just construct it in the load function

**Review by ls-1801 — CHANGES_REQUESTED** [review]

File: .github/workflows/build_and_test.yml (hunk 6 above)
Comment (keyseven123) [reply .github/workflows/build_and_test.yml]: Yes, sorry. I forgot to drop the commit before re-requesting feedback

File: nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp (hunk 11 above)
Comment (keyseven123) [reply nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp]: Only query compilation time, but I think we can do this at a later stage, if the compilation time becomes a problem

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp (hunk 13 above)
Comment (keyseven123) [reply nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: Unfortuantely, it is not possible to pass a double to an `inRange()`. I have added a workaround. Let me know what you think about it.

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp (hunk 14 above)
Comment (keyseven123) [reply nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: Done

File: nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp (hunk 10 above)
Comment (keyseven123) [reply nes-nautilus/src/Interface/Hash/BloomFilterRef.cpp]: Yes sir.

File: nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp (hunk 7 above)
Comment (keyseven123) [reply nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp]: ChainedHashMap has now an optional

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp (hunk 8 above)
Comment (keyseven123) [reply nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp]: Added a `ChainedHashMapConfig` object

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp (hunk 9 above)
Comment (keyseven123) [reply nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp]: Do not need this anymore

File: nes-physical-operators/include/HashMapOptions.hpp (hunk 15 above)
Comment (keyseven123) [reply nes-physical-operators/include/HashMapOptions.hpp]: It is now an optional BloomFilterParams.

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp (hunk 12 above)
Comment (keyseven123) [reply nes-nautilus/tests/UnitTests/BloomFilterTest.cpp:288]: Changed.

File: nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    /// size the `ChainedHashMap`'s in-map bit area so disabled maps allocate nothing.
+    /// Rounded up to whole 64-bit words: `add`/`mightContain` index by word (`bitCount - 1` may land in the
+    /// last, partially used word), so a byte-exact size would leave that word's tail out of bounds.
+    [[nodiscard]] uint64_t allocationByteCount() const
+    {
+        if (hashCount == 0)
+        {
+            return 0;
+        }
+        return ((bitCount + 63) / 64) * sizeof(uint64_t);
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp]: Could we validate the Bloom-filter input parameters at their construction boundaries? In practice, a `bitCount` above `UINT64_MAX - 63` is already invalid, but because `BloomFilterParams` is aggregate-constructible it can make this rounding overflow into an undersized allocation; similarly, an enabled filter with `bitCount == 0` later performs modulo by zero. `compute()` should also require a finite `fpRate` strictly between 0 and 1: negative/NaN values can reach a NaN-to-`uint64_t` conversion, while zero can derive an enormous `hashCount` for Nautilus to statically unroll. A precondition or checked factory would make these contracts explicit.

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp [hunk 17]
```diff
@@ -142,12 +173,23 @@ class ChainedHashMap final : public HashMap
         uint64_t mask;
         ChildBufferIndex storageSpaceIndex;
         ChildBufferIndex varSizedSpaceIndex;
+        /// Sizing of the optional in-map BloomFilter, nullopt when disabled. The bit area itself sits inline
+        /// behind the chains array, so no index into a child buffer is needed.
+        std::optional<Nautilus::Interface::BloomFilterParams> bloomFilter;
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp:178]: Maybe as a follow-up: Do we need to store the Bloom-filter configuration in both the hash-map header and the view? For example, `entrySize` is stored in the header even though it is also passed to `ChainedHashMapRef` when the view is created. The Bloom-filter configuration is now duplicated in the same way. Could the view be the source of these constant values, while the header stores only state that changes at runtime? This may also apply to our other data structures and does not need to block this PR.

File: nes-query-compiler/src/LoweringRules/LowerToPhysical/LowerToPhysicalHashJoin.cpp [hunk 18]
```diff
@@ -255,7 +256,9 @@ HashMapOptions createHashMapOptions(
         keySize,
         valueSize,
         pageSize,
-        numberOfBuckets};
+        numberOfBuckets,
+        /// HashJoin does not yet enable the ChainedHashMap's optional filter.
+        std::nullopt};
```
Comment (ls-1801) [inline nes-query-compiler/src/LoweringRules/LowerToPhysical/LowerToPhysicalHashJoin.cpp]: Why is the Bloom filter not enabled for hash joins in general here? We probably want a configuration such as `worker.default_query_execution.bloom_filter_in_hash_join` to turn it off for debugging. To cover both modes without duplicating the complete system-test suite, we could add `worker.default_query_execution.bloom_filter_in_hash_join: [true, false]` to the configuration matrix in one of the existing hash-join system-test files.

**Review by ls-1801 — CHANGES_REQUESTED** [review]

File: nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp (hunk 16 above)
Comment (keyseven123) [reply nes-nautilus/include/Interface/Hash/BloomFilterRef.hpp]: Done

File: nes-query-compiler/src/LoweringRules/LowerToPhysical/LowerToPhysicalHashJoin.cpp (hunk 18 above)
Comment (keyseven123) [reply nes-query-compiler/src/LoweringRules/LowerToPhysical/LowerToPhysicalHashJoin.cpp]: Because I suck...

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp (hunk 17 above)
Comment (keyseven123) [reply nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp:178]: I have created #1856 and #1858 for this.

File: nes-nautilus/tests/UnitTests/BloomFilterTest.cpp [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+        RC_ASSERT(actualFpRate <= tolerance);
+    }
+    else
+    {
+        NES_INFO("Skipping FP rate check: only {} true negative samples (need {})", trueNegatives.size(), MIN_FP_SAMPLES_FOR_RATE_CHECK);
+    }
+}
+
+/// Property: zeroing the bit array reverts every previously-inserted key to mightContain=false.
+/// Single-threaded so the clear is uncontested; the multi-threaded race surface is covered by the
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/BloomFilterTest.cpp]: This says the main property covers the multi-threaded race surface, but `insertAndCheckProperty` is entirely single-threaded (and even documents that at line 159). Please remove this claim or add an actual concurrent test if concurrent access is part of the contract.

**Review by ls-1801 — APPROVED** [review]
just one comment fixup
